sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosException.java

```diff
@@ -299,10 +314,39 @@ public double getRequestCharge() {
 
     @Override
     public String toString() {
-        return getClass().getSimpleName() + "{" + "userAgent=" + USER_AGENT + ", error=" + cosmosError + ", resourceAddress='"
-                   + resourceAddress  +  ", statusCode=" + statusCode + ", message=" + getMessage()
-                   + ", causeInfo=" + causeInfo() + ", responseHeaders=" + responseHeaders + ", requestHeaders="
-                   + filterSensitiveData(requestHeaders) + '}';
+        try {
+            ObjectNode exceptionMessageNode = mapper.createObjectNode();
+            exceptionMessageNode.put("ClassName", getClass().getSimpleName());
+            exceptionMessageNode.put(USER_AGENT_KEY, USER_AGENT);
+            exceptionMessageNode.put("statusCode", statusCode);
+            exceptionMessageNode.put("resourceAddress", resourceAddress);
+            if (cosmosError != null) {
+                exceptionMessageNode.put("error", cosmosError.toJson());
+            }
+
+            exceptionMessageNode.put("innerErrorMessage", innerErrorMessage());
+            exceptionMessageNode.put("causeInfo", causeInfo());
+            if (responseHeaders != null) {
+                exceptionMessageNode.put("responseHeaders", responseHeaders.toString());
+            }
+
+            List<Map.Entry<String, String>> filterRequestHeaders = filterSensitiveData(requestHeaders);
+            if (filterRequestHeaders != null) {
+                exceptionMessageNode.put("requestHeaders", filterRequestHeaders.toString());
+            }
+
+            if(this.cosmosDiagnostics != null) {
+                cosmosDiagnostics.fillCosmosDiagnostics(exceptionMessageNode, null);
```

Comment:
serializing diagnostics is very expensive. Now that we are doing that in the exception stacktrace for each exception we should exclude the buissness logic errors: NotFound (with no substatus code), Conflict, PreconditionFailed.
These are buissness logic failures we don't need to the diagnostics for this

Reply:
Before this PR we were still serializing diagnostics and putting in string on exception message. In this PR we putting diagnostics object in exception parent ObjectNode  and then serializing in json at the last. So for computation on diagnostics serialization still remain the same.
However I see your point in general to avoid diagnostics serialization on regular customer business scenario like 404 not found. I would like to keep it as sperate work item, however there is down side to put custom logic based on status code, we need to maintain extra logic for future scenarios and secondly its customer who should decide  whether to fetch the whole message when there is expected failure and can be handle based on expected status code.

Reply:
We should exclude the business logic failure (404, 409, 412) and not include in the exception due to the perf impact.

Feel free to capture this work in a separate PR.

Reply:
https://github.com/Azure/azure-sdk-for-java/issues/18271